### ISSUE TITLE
[Bug] IE11 fails to show chart data after a few minutes

### ISSUE THREAD

**Issue by atulc** [grafana/grafana#6936]
Please prefix your title with [Bug] or [Feature request]
For question please check [Support Options](http://grafana.org/support/). **Do not** open a github issue

Please include this information:
- Version: Grafana 4.0.1
- Datasource: Influxdb
- OS: Windows 7 & 10
- What was I doing?  Just watching a chart over time (every 5s refresh)
- What should happen? Data continues to appear
- What happened instead?   Data disappears

If I do this with Chrome it's OK though.
![ie11-windows10-hdpi-no-data](https://cloud.githubusercontent.com/assets/2449316/21115349/f890e00a-c07e-11e6-975d-ef0d0fa522c5.png)
![ok-chrome-windows7](https://cloud.githubusercontent.com/assets/2449316/21115350/f892fffc-c07e-11e6-9748-57b8f9d91cde.png)

 

**IMPORTANT** 
If it relates to *metric data viz*:
- An image or text representation of your metric query
- The raw query and response for the network request (check this in chrome dev tools network tab, here you can see metric requests and other request, please include the request body and request response)

If it relates to *alerting*
- An image of the test execution data fully expanded.

**IMPORTANT** 
If it relates to *metric data viz*:
- An image or text representation of your metric query
- The raw query and response for the network request (check this in chrome dev tools network tab, here you can see metric requests and other request, please include the request body and request response)

If it relates to *alerting*
- An image of the test execution data fully expanded.

**Comment by torkelo** [comment]
this only happens on IE11 ? Checked the developer tools network log? what are the metric requests & responses?

**Comment by wregen** [comment]
I can reproduce the bug. In my case it appears in IE11 and Edge on Windows 10. The behavior of the graph is identical to the one described by @atulc. The graph is rendered once the user enters the dashboard and it slowly disappears (moving to the left). The data source is InfluxDb v1.1.1. accessed by the Grafana 4.0.1 client through the proxy (Http Settings: Access: proxy). The Time range is set to "Last 5 minutes" and "Refresh every:"  is set to 5s.

IE11 cache settings: Go to "Tools" -> "Internet Options" -> "General Tab, Browsing History section". Click "Settings" button. Then on "Website Data Settings" dialog check "Check for newer versions of strored pages" options to "Automatically". With such settings IE11 requests the data from InfluxDb only once.

The query that the "graph reissues" every 5 seconds in my case looks as follows:
http://192.168.1.119:3000/api/datasources/proxy/1/query?db=telegraf&q=SELECT%20value%20FROM%20%22mymeasurement%22%20WHERE%20%22sensor%22%20%3D%20%27Voltage%20L1-N%27%20AND%20time%20%3E%20now()%20-%205m%20fill(null)%3BSELECT%20mean(%22value%22)%20FROM%20%22mymeasurement%22%20WHERE%20%22sensor%22%20%3D%20%27Voltage%20L1-N%27%20AND%20time%20%3E%20now()%20-%205m%20GROUP%20BY%20time(10s)%20fill(null)%3BSELECT%20last(%22value%22)%20FROM%20%22mymeasurement%22%20WHERE%20%22sensor%22%20%3D%20%27Voltage%20L1-N%27%20AND%20time%20%3E%20now()%20-%205m&epoch=ms

As you can see on the screenshot below with such IE11 settings, Explorer does not get the data from the server as - due to the fact that the query (URL) is not changing - it assumes that the data is not changed. (See the Receiver column on developer console at the bottom of the screenshot.)

The behavior changes to correct one if in cache settings I change the "Check for newer versions of stored pages" to "Every time I visit the page" (which I treat as a workaround). But with such settings no data is cached which slows down the browsing in general.

The problem is well known, IE (even 11) caches GET AJAX requests. And there are several techniques to workaround it:
1) Using a unique query string parameter (such as a unix timestamp) on each request, so as to make each request URL unique, thereby preventing caching.
2) using POST instead of GET, as IE does not cache POST requests in normal circumstances.
3) using 'cache-control' headers passed by the server.

I am not sure how to fix it, as I understand that Grafana is working as a proxy here for InfluxDb HTTP interface and InfluxDb does not send any 'cache-control' headers in response.

![screenshot - 2017-01-10 - 03 59 03](https://cloud.githubusercontent.com/assets/64548/21793053/6a30a026-d6ee-11e6-9c11-f8a0702bb28b.png)

**Comment by wregen** [comment]
Hi,
I've made the following change against v4.0.1 (and recompiled the grafana-server) which fixes the problem of disappearing graphs in IE11 and Edge (when using grafana with influxdb as data-source).
I am not sure however if the solution with extra headers is general enough for other data-sources. 

```
diff --git a/pkg/api/dataproxy.go b/pkg/api/dataproxy.go
index 39de691..4131c44 100644
--- a/pkg/api/dataproxy.go
+++ b/pkg/api/dataproxy.go
@@ -194,6 +194,14 @@ func ProxyDataSourceRequest(c *middleware.Context) {
                c.JsonApiErr(400, "Unable to load TLS certificate", err)
                return
        }
-       proxy.ServeHTTP(c.Resp, c.Req.Request)
-       c.Resp.Header().Del("Set-Cookie")
+
+       c.Resp.Header().Set("Expires", "0")
+       c.Resp.Header().Set("Cache-control", "no-cache")
+       c.Resp.Header().Set("Cache-control", "no-store")
+       c.Resp.Header().Set("Pragma", "no-cache")
+
+        proxy.ServeHTTP(c.Resp, c.Req.Request)
+        c.Resp.Header().Del("Set-Cookie")
+
 }
```

**Comment by HeikoOnnebrink** [comment]
Version: Grafana 4.1.2
Datasource: Influxdb 0.13
OS: Docker Container build from ubuntu:14.04 

What was I doing? Just watching a chart over time (every 30s refresh)
What should happen? Data continues to appear
What happened instead? Data disappears

Hi,
just ran into same issue as described above for clients using iExplorer 11 when using Grafana 3.1
Edge and Firefox all work fine, just in iE11 data does not get rendered into the graphs during page refreshes while initial page load displays all graph data.

As you mentioned that issue should be fixed in 4.0.1 
I build new docker container based on Grafana 4.1.2 (as I assume that fix intoruced in 4.0.1 made its way into 4.1.2). But still same problem

The workaround for iE11 to change cache settings from "Automatic" to "Every time I visit the page" fixes the problem after a page reload, but no need to discuss that this is no permanent solution.

Funny to mention that I never could reproduce the problem on Edge while you reported that you were able to reproduce it also there. I even did not find any setting inside Edge to change cache behavior (apart from the buttons in developer tools that are only active as long as developer tools are open ...)
